src/Microsoft.ML.Core/Utilities/Contracts.cs

```diff
@@ -454,14 +455,22 @@ public static Exception ExceptSchemaMismatch(this IExceptionContext ctx, string
             => Process(new ArgumentOutOfRangeException(paramName, MakeSchemaMismatchMsg(columnRole, columnName)), ctx);
         public static Exception ExceptSchemaMismatch(string paramName, string columnRole, string columnName, string expectedType, string actualType)
             => Process(new ArgumentOutOfRangeException(paramName, MakeSchemaMismatchMsg(columnRole, columnName, expectedType, actualType)));
+        public static Exception ExceptSchemaMismatch(string paramName, string columnRole, string columnName, Type expectedType, string actualType)
+            => Process(new ArgumentOutOfRangeException(paramName, MakeSchemaMismatchMsg(columnRole, columnName, expectedType.ToString(), actualType)));
+        public static Exception ExceptSchemaMismatch(string paramName, string columnRole, string columnName, Type[] expectedType, string actualType)
```

Comment:
I do not like this very much. First of all it is using the wrong type... the `IDataView` type system is based on its own type system, which is distinct from the .NET type system. (E.g., key values and unsigned integers have the same .NET type, but different `ColumnType` instances.) Also one of the most common instances of this is something abstract like, "a known sized vector of float," which is not a concept expressible in this. Also it seems to be introducing a "convenience" for a very specialized case.

So, all this is to say, this last addition I am unenthusiastic about. Could we revert that please, and reconsider?

Reply:
That sounds good, will do. What do you think of the use of typeof(float).ToString() instead of using "float"?

Reply:
After discussing with Tom, I am reverting my changes only keeping the switch from "input" to "source" column in transformers. @singlis

Reply:
Thanks @artidoro . I think I should probably explain my thinking here, since it may seem counterintuitive.

So generally I've found that having conveniences to make things more consistent is a good thing. However, one thing that I've observed, even in this PR, is that for ***exception messages specifically***, I am a little wary of it. I consider the problems of #2185. One of the problems is the error messages are wrong, but to a casual reader it is not *obvious* that the error messages are wrong, since it is buried behind layers of "convenience" to make the messages consistent. I don't think the original author had ill intent, but they just got confused about what "convenience" led to what message and what the meaning of the inputs is. (As in fact we also saw here, with the controversy about "label" vs. "Label", and whether something was supposed to be a column name or not!!)

I'm not sure how to solve that issue, but it seems to me like this `ExceptSchemaMismatch` has some problem that makes it badly conceived if its effect of "consistency" is that people are writing wrong error messages, and being fundamentally confused about it. I don't know how to solve that issue. But the change to introduce *more* layers of abstraction would only compound the problem that these "conveniences" are hard to use and people evidently find them very confusing.

This is not to say I think we should not use `ExceptSchemaMismatch`, but I think it needs to be clarified in some fashion. Right now to a reviewer it is really, really hard to tell what is a "right" vs. "wrong" message (as seen by all the discussions in this PR), and the suggestion to get out of that hole by continuing to dig was not one I liked very much. Which is why I had to disagree with some other suggestions in this PR, sorry.